azure-iot-device/azure/iot/device/common/pipeline/operation_flow.py

```diff
@@ -15,124 +15,145 @@
-    :param PipelineOperation new_op: Operation that is being passed down the pipeline
-      to effectively continue the work represented by original_op.  This is most likely
-      a different type of operation that is able to accomplish the intention of the
-      original_op in a way that is more specific than the original_op.
-    """
-
-    logger.debug("{}({}): continuing with {} op".format(stage.name, original_op.name, new_op.name))
+class PipelineFlow(object):
+    @pipeline_thread.runs_on_pipeline_thread
+    def _send_worker_op_down(self, worker_op, op):
+        """
+        Continue an operation using a new operation.  This means that the new operation
+        will be passed down the pipeline (starting at the next stage). When that new
+        operation completes, the original operation will be completed.  In this way,
+        a stage can accept one type of operation and, effectively, change that operation
+        into a different type of operation before passing it to the next stage.
+
+        This is useful when a generic operation (such as "enable feature") needs to be
+        converted into a more specific operation (such as "subscribe to mqtt topic").
+        In that case, a stage's _execute_op function would call this function passing in
+        the original "enable feature" op and the new "subscribe to mqtt topic"
+        op.  This function will pass the "subscribe" down. When the "subscribe" op
+        is completed, this function will cause the original op to complete.
+
+        This function is only really useful if there is no data returned in the
+        new_op that that needs to be copied back into the original_op before
+        completing it.  If data needs to be copied this way, some other method needs
+        to be used.  (or a "copy data back" function needs to be added to this function
+        as an optional parameter.)
+
+        :param PipelineOperation op: Operation that is being continued using a
+          different op.  This is most likely the operation that is currently being handled
+          by the stage.  This operation is not actually continued, in that it is not
+          actually passed down the pipeline.  Instead, the original_op operation is
+          effectively paused while we wait for the new_op operation to complete.  When
+          the new_op operation completes, the original_op operation will also be completed.
+        :param PipelineOperation worker_op: Operation that is being passed down the pipeline
+          to effectively continue the work represented by the original op.  This is most likely
+          a different type of operation that is able to accomplish the intention of the
+          original op in a way that is more specific than the original op.
+        """
+
+        logger.debug("{}({}): continuing with {} op".format(self.name, op.name, worker_op.name))
+
+        @pipeline_thread.runs_on_pipeline_thread
+        def worker_op_complete(new_op, error):
```

Comment:
Also I'm not sure if "error" is the right semantics here for this arg. We've discussed this before, but I think it's more relevant now - are these truly errors? After all, especially as we get more and more robust in the pipeline, we're definitely throwing exceptions around that are recoverable. Retry for instance.

That being said, "error" is referring to the status of the operation itself, so, I'm in favor of there being no change here, provided that an exception (even a recoverable one) that is attached to an op, always indicates failure of an op.

That is to say:

1) An exception occurs which can be recovered from (e.g. can't connect, will retry)
2) This fails the current operation
3) Another operation is spawned in order to recover (e.g. retry the connect again)

If that's always true, then I still consider this to be an error in the context of an operation. However if an operation can have an "error" set, that is recoverable without failure completion of operation, then I think the semantics are wrong.

That all said, it seems to me that everything is okay, and the semantics ARE correct. Just wanted to double check that my interpretation of how this is supposed to work is all correct.